### PR TITLE
add submarine

### PR DESCRIPTION
**Pull request by damonhuang233**

### CODE REVIEW

**Comment by 1mgroot** [comment]
Looks good.

File: src/main/java/cs361/battleships/models/Board.java [hunk 1]
```diff
@@ -32,15 +33,35 @@ public Board() {
...
 			return false;
 		}
 		if (ships.stream().anyMatch(s -> s.getKind().equals(ship.getKind()))) {
 			return false;
 		}
 		final var placedShip = new Ship(ship.getKind());
 		placedShip.place(y, x, isVertical);
-		if (ships.stream().anyMatch(s -> s.overlaps(placedShip))) {
+
+		if(placedShip.getKind().equals("SUBMARINE")){
```
Comment (kryolight) [inline src/main/java/cs361/battleships/models/Board.java:46]: This place function allows submarines to be placed in the same square as other ships after the other ships are placed. To reduce the amount of networking routes we need to make. The function infers that the player wants the submarine to be submerged when the player tries to place the submarine in an overlapping square. In all other cases the submarine is treated as a surface ship.

File: src/main/java/cs361/battleships/models/Board.java [hunk 2]
```diff
@@ -66,6 +87,45 @@ public Result attack(int x, char y) {
 	}
 
 	private Result attack(Square s) {
+
```
Comment (kryolight) [inline src/main/java/cs361/battleships/models/Board.java:90]: The attack function only attacks one ship in a square at a time. If two ships are located at the same location it takes two shots to hit both ships. The function checks if submerged subs can be hit before it scores hits on them.

**Review by kryolight — COMMENTED** [review]
This pull request includes code for the submarine including the submarine placing function, updated attack function, and helper functions. I'm concerned with how the move command interacts with submarines. It seems that submarines will move unhindered by other ships and vice versa. This means that a submarine can go from being in a submerged state to an unsubmerged state and vice versa. The code to works, however it doesn't follow the specifications of the new features completely, instead it takes a twist following our own unique specifications.

**Review by damonhuang233 — COMMENTED** [review]
Need more test later

**Comment by kryolight** [comment]
All the code except the attack code meets the assignment's specifications. The attack code will be fixed with issue #59 .

**Pull request merged**